### PR TITLE
SPARK-8153 Add configuration for disabling partial aggregation in runtime

### PR DESCRIPTION
**Pull request by navis**
Same thing with "hive.map.aggr.hash.min.reduction" in hive, which disables hash aggregation if it's not sufficiently decreasing the output size.

Added two configuration
- spark.sql.partial.aggregation.checkInterval
- spark.sql.partial.aggregation.minReduction

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

**Comment by navis** [comment]
Test fail was just caused by appearance order. Added order-by for deterministic result

File: core/src/main/scala/org/apache/spark/util/collection/BitSet.scala [hunk 1]
```diff
@@ -32,6 +34,10 @@ class BitSet(numBits: Int) extends Serializable {
    */
   def capacity: Int = numWords * 64
 
+  def clear(): Unit = {
+    util.Arrays.fill(words, 0L)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/collection/BitSet.scala]: Use Scala classes in Scala code

Comment (navis) [inline core/src/main/scala/org/apache/spark/util/collection/BitSet.scala]: I'm new to scala. Could you suggest one?

Comment (srowen) [inline core/src/main/scala/org/apache/spark/util/collection/BitSet.scala]: Hm, sorry, I take that back. There is no equivalent in Scala that I can find, that sets an existing array's values. Can I suggest thought that importing just `java.util` might look confusing, so either just import it (there is no  Scala `Arrays` to mix it up with) or if this is just one usage, write `java.util.Arrays.fill(...)`

Comment (navis) [inline core/src/main/scala/org/apache/spark/util/collection/BitSet.scala]: Addressed comments. Thanks!

**Comment by JoshRosen** [comment]
It looks like this is failing its own test.

**Comment by navis** [comment]
Strange.. cannot reproduce the fail in local env. I'll check it again.

**Comment by navis** [comment]
The memory leakage caused the test fail was a existing bug in master branch. Currently, unsafe-based hash is released on 'next' call but if input is empty, it would not be called ever. 
Now returns an empty iterator if input is empty.

**Comment by JoshRosen** [comment]
@navis, good catch on finding the memory leak in the unsafe aggregation path.  I think that maybe we should extract that bugfix into its own PR so that it's easier to backport to 1.4.x.

**Comment by navis** [comment]
@JoshRosen ok, sure.

**Comment by navis** [comment]
done in SPARK-8357

**Comment by andrewor14** [comment]
A note about the naming: I think you meant `spark.sql.partialAggregation.*` instead? It doesn't make sense to have a `spark.sql.partial.*` namespace.

**Comment by andrewor14** [comment]
@JoshRosen @marmbrus any updates on the details of this patch?

**Comment by marmbrus** [comment]
Unfortunately, the Aggregate1 code path is deprecated and going to be removed shortly.  We should probably close this issue and design the feature on JIRA for the new aggregation code path.

**Pull request closed without merging**